### PR TITLE
Feature/attribute succession

### PR DESCRIPTION
**Pull request by slavistan**

### CODE REVIEW

File: CHANGELOG.md
Comment (jkoenig134) [inline CHANGELOG.md:8]: maybe list the added routes?

File: CHANGELOG.md
Comment (jkoenig134) [inline CHANGELOG.md:7]:
```suggestion
-   fix: `POST /api/v2/Attributes` will no longer accept RelationshipAttributes, as this was unintentional and led to wrong behavior
```

File: packages/sdk/src/endpoints/AttributesEndpoint.ts
Comment (jkoenig134) [inline packages/sdk/src/endpoints/AttributesEndpoint.ts:5]: why is this aliased (vs renamed in the actual file)?

**Comment by slavistan** [comment]
Local tests run fine on mongo db (`npm run test:local:mongo`), but fail when run on ferret db (`npm run test:local:ferret`).

**Comment by jkoenig134** [comment]
@slavistan same with ci tests. Let's first tackle the review and then look at what could be the root cause.

File: src/modules/coreHttpApi/controllers/AttributesController.ts
Comment (jkoenig134) [inline src/modules/coreHttpApi/controllers/AttributesController.ts:20]: I think to be really backwards compatible this check should be re-added. I mean - if owner if specified AND is set to not the own address we just delete it.

File: test/attributes.test.ts
Comment (jkoenig134) [inline test/attributes.test.ts:70]: this is copy paste from the first test (`should create a repository attribute` lowercase)

File: test/iqlquery.test.ts
Comment (jkoenig134) [inline test/iqlquery.test.ts:78]: why isn't this optional anymore?

File: test/lib/testUtils.ts
Comment (jkoenig134) [inline test/lib/testUtils.ts:67]: missing newline

File: test/attributes.test.ts
Comment (jkoenig134) [inline test/attributes.test.ts:279]: I don't think we need to duplicate the query here. Querying at one connector completely satisfies this test.

File: test/attributes.test.ts
Comment (jkoenig134) [inline test/attributes.test.ts:72]: I'd like to see that to peer is actually notified here

File: test/attributes.test.ts
Comment (jkoenig134) [inline test/attributes.test.ts:101]: I'd like to see that the attribute is actually succeeded here

File: test/attributes.test.ts
Comment (jkoenig134) [inline test/attributes.test.ts:127]: see above

**Review by jkoenig134 — APPROVED** [review]

**Comment by jkoenig134** [comment]
It seems like the tests fail against ferretdb when ferretdb is running against postgres. This is REALLY bad, but shouldn't block us for now.

File: test/attributes.test.ts
Comment (Magnus-Kuhn) [inline test/attributes.test.ts:95]: consistent variable naming throughout the tests: e. g. succeedAttribute returns a successionResponse, a succeedAttributeResponse and a result, notifyPeerAbout... returns a notificationResult

File: test/iqlquery.test.ts
Comment (sebbi08) [reply test/iqlquery.test.ts:78]: @J-Gann do you know that ?

**Review by jkoenig134 — CHANGES_REQUESTED** [review]

**Review by jkoenig134 — DISMISSED** [review]
LGTM, @stnmtz please also leave a review.

**Comment by stnmtz** [comment]
Found some things, I'm on it reviewing

File: test/lib/customMatcher.ts
Comment (stnmtz) [inline test/lib/customMatcher.ts:16]: exclude?

File: test/lib/customMatcher.ts
Comment (stnmtz) [inline test/lib/customMatcher.ts:51]: export {} ? Why is that? I would for a comment.

File: src/modules/coreHttpApi/openapi.yml
Comment (stnmtz) [inline src/modules/coreHttpApi/openapi.yml:547]: Own shared identity attributes is mentioned here, would it be better to use repository attributes?

File: src/modules/coreHttpApi/openapi.yml
Comment (stnmtz) [inline src/modules/coreHttpApi/openapi.yml:482]: > Succeeding a Relationship Attribute will return an

Use present tense. predecessor/successor is missing, why "additional" ID?

File: src/modules/coreHttpApi/openapi.yml
Comment (stnmtz) [inline src/modules/coreHttpApi/openapi.yml:448]: Casing of Repository Attribute vs. repository attribute below

File: src/modules/coreHttpApi/openapi.yml
Comment (stnmtz) [inline src/modules/coreHttpApi/openapi.yml:480]: Why 200 on POST?

File: src/modules/coreHttpApi/controllers/AttributesController.ts
Comment (stnmtz) [inline src/modules/coreHttpApi/controllers/AttributesController.ts:22]: cannotCreateNotSelfOwnedAttribute => cannotCreateNotSelfOwnedRepositoryAttribute

File: packages/sdk/src/types/attributes/IQLQuery.ts
Comment (stnmtz) [inline packages/sdk/src/types/attributes/IQLQuery.ts:2]: As Julian asked in the tests, why no longer optional?

File: README_dev.md
Comment (stnmtz) [inline README_dev.md:7]: Why optional? Why "make sure to install it ..." when it is optional?

**Review by stnmtz — CHANGES_REQUESTED** [review]

File: README_dev.md
Comment (sebbi08) [reply README_dev.md:7]: I think it is optional as it is only required for manual restart. By default, the restart will be handled by nodemon

Comment (jkoenig134) [reply README_dev.md:7]: I mean, partly true: nodemon watches the files and restarts the server on changes. But sometimes (e.g. when nodemon fails to startup the server) we need to have a manual way to do it.

But yeah, it's very optional b/c it's only needed soooooometimes :D

File: test/lib/customMatcher.ts
Comment (jkoenig134) [reply test/lib/customMatcher.ts:51]: this is for typescript to actually compile this. You instantly get a compiler error when removed so I vote against this comment :)

File: src/modules/coreHttpApi/openapi.yml
Comment (sebbi08) [reply src/modules/coreHttpApi/openapi.yml:482]: The ID is only set when a relationship attribute is suceeded as this will directly be send to the peer via notification

File: test/lib/customMatcher.ts
Comment (sebbi08) [reply test/lib/customMatcher.ts:51]: actually not needed anymore seems to be an update in ts or something else :/

File: packages/sdk/src/types/attributes/requests/CreateIdentityAttributeRequest.ts
Comment (sebbi08) [reply packages/sdk/src/types/attributes/requests/CreateIdentityAttributeRequest.ts:3]: done

File: README_dev.md
Comment (slavistan) [inline README_dev.md:5]: What's the point in installing `typescript` and `cpx` globally? Let's add `typescript` and `cpx` as dev dependencies instead.

Comment (jkoenig134) [reply README_dev.md:5]: cpx had vulnerabilities for a while (edit: still has) and wasn't included in the repo therefore **(leave this as is!)** 

typescript is already included, but WAY MORE better to use when installed globally. If you don't want to have it globally installed just skippit :)

Comment (sebbi08) [reply README_dev.md:5]: We could use https://github.com/bcomnes/cpx2 as this is maintained and a drop-in replacement

Comment (jkoenig134) [reply README_dev.md:5]: Ok, sounds great!

Comment (jkoenig134) [reply README_dev.md:5]: as this is not part of this PR I'll add this in another PR.

File: packages/sdk/src/types/attributes/IQLQuery.ts
Comment (sebbi08) [reply packages/sdk/src/types/attributes/IQLQuery.ts:2]: Hi this is a little bit irritating for me.

https://github.com/nmshd/runtime/blob/39b5138435a9b7e137a36d21278c76fa81e0fe33/packages/content/src/attributes/IQLQuery.ts#L33

Here it also is not optional anymore. But in the `from` method the property is omitted

https://github.com/nmshd/runtime/blob/39b5138435a9b7e137a36d21278c76fa81e0fe33/packages/content/src/attributes/IQLQuery.ts#L59

Comment (sebbi08) [reply packages/sdk/src/types/attributes/IQLQuery.ts:2]: When I try to call the runtime use case and omit the `"@type": "IQLQuery"` i will get an error that it could not be parsed

File: src/modules/coreHttpApi/openapi.yml
Comment (stnmtz) [inline src/modules/coreHttpApi/openapi.yml:545]: Again 200 on POST :P

File: packages/sdk/src/types/attributes/IQLQuery.ts
Comment (stnmtz) [reply packages/sdk/src/types/attributes/IQLQuery.ts:2]: > When I try to call the runtime use case and omit the "@type": "IQLQuery" i will get an error that it could not be parsed
Ok, that's strange ...

- As we directly deserialize via IQLQuery.from with the Runtime's Usecase (see https://github.com/nmshd/runtime/blob/main/packages/runtime/src/useCases/consumption/attributes/ExecuteIQLQuery.ts#L28 ) we do not need @type: "IQLQuery" within the JSON for any payload to POST .../ExecuteIQLQuery (as ts-serval is automatically using the default type then)
- I guess this is caused by a TypeScript type error b/c the interface of ExecuteIQLQueryRequest seems to be wrong, as it should be possible to omit the @ type definition in the request (because of the first bullet point): https://github.com/nmshd/runtime/blob/main/packages/runtime/src/useCases/consumption/attributes/ExecuteIQLQuery.ts#L10
- I can remember that @slavistan had some problems when building the schemas within the Runtime because of some ugly bugs which might be a reason/cause of not omitting the @ type within the ExecuteIQLQueryRequest interface in the Runtime.
- BUT: I can imagine that using the Connector however all works out good, as all the requests coming in are any's and thus the TypeScript error is ommitted.
- Thus: I can imagine that using the SDK without the @ type: "IQLQuery" definition is working great, and we could remove this change. But we still have an inconsistency within the Runtime's request types of the Usecases.


Furthermore:
- Compared to RelationshipAttributeQuery, we're doing exact the same thing within the Runtime (the @ type is defined within the RelationshipAttributeQuery): https://github.com/nmshd/runtime/blob/39b5138435a9b7e137a36d21278c76fa81e0fe33/packages/content/src/attributes/RelationshipAttributeQuery.ts#L69
- BUT within the type definition of RelationshipAttributeQueryJSON in the Connector SDK, the @ type is optional: https://github.com/nmshd/connector/blob/main/packages/sdk/src/types/attributes/RelationshipAttributeQuery.ts#L2
- So as we're doing the same thing, using the Runtime Usecase ExecuteRelationshipAttributeQuery without an @ type:"RelationshipAttributeQuery" is also doing fine?
- And we might have the same Runtime request type issues for Execute... (or even more?)

Comment (J-Gann) [reply packages/sdk/src/types/attributes/IQLQuery.ts:2]: I believe the error occurs during validation of the UseCase: https://github.com/nmshd/runtime/blob/main/packages/runtime/src/useCases/consumption/attributes/ExecuteIQLQuery.ts#L15, as the schema of "ExecuteIQLQueryRequest" requires @type to be defined:
https://github.com/nmshd/runtime/blob/main/packages/runtime/src/useCases/common/Schemas.ts#L15948
This validation seems to be missing in the IdentityAttributeQuery: https://github.com/nmshd/runtime/blob/main/packages/runtime/src/useCases/consumption/attributes/ExecuteIdentityAttributeQuery.ts#L15

Comment (jkoenig134) [reply packages/sdk/src/types/attributes/IQLQuery.ts:2]: I'll wait for this to be merged and will update the code

Comment (sebbi08) [reply packages/sdk/src/types/attributes/IQLQuery.ts]: is optional again

Comment (jkoenig134) [reply packages/sdk/src/types/attributes/IQLQuery.ts:2]: I've gone a quite different route - recreating what we have in the actual types. Now we can omit it when KNOWING what kind of query we have.

**Review by jkoenig134 — DISMISSED** [review]
LGTM, but please wait for @stnmtz to also approve.

**Review by jkoenig134 — DISMISSED** [review]

**Comment by jkoenig134** [comment]
@stnmtz we will merge this, if you find sth. we can do this afterwards.

**Pull request merged**
